field_audit/field_audit.py

```diff
@@ -0,0 +1,104 @@
+import contextvars
+from functools import wraps
+
+from django.db import models
+
+from .utils import get_fqcn
+
+__all__ = [
+    "AlreadyAudited",
+    "audit_fields",
+    "get_audited_class_path",
+    "get_audited_models",
+    "request",
+]
+
+
+class AlreadyAudited(Exception):
+    """Class is already audited."""
+
+
+def audit_fields(*field_names, class_path=None):
+    """Class decorator for auditing field changes on DB model instances.
+
+    Use this on Model subclasses which need field change auditing.
+
+    :param field_names: names of fields on the model that need to be audited
+    :param class_path: optional name to use as the ``object_class_path`` field
+        on audit events. The default (``None``) means the audited model's
+        fully qualified "dot path" will be used.
+    """
+    def wrapper(cls):
+        if cls in _audited_models:
+            raise AlreadyAudited(cls)
+        if not issubclass(cls, models.Model):
+            raise ValueError(f"expected Model subclass, got: {cls}")
+        cls.__init__ = _decorate_init(cls.__init__, field_names)
+        cls.save = _decorate_db_write(cls.save, field_names)
+        cls.delete = _decorate_db_write(cls.delete, field_names)
```

Comment:
not blocking (and I can't recall if this was already discussed) but can you say why you chose to wrap the methods instead of using Django signals?

Reply:
Sure. One of the reasons is because `_decorate_db_write` needs to perform logic before and after the write method is called, and using signals would have required splitting that logic into separate signal listeners (`pre_<event>` and `post_<event>`). In other words this really is use case for decorators because Django's signals framework would have fragmented the logic and made this code significantly more complex.